Towards supporting quantized structured kernels

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #72353
* __->__ #74560

This PR add support for quantized tensors with "unknown quantizer",
which means that we can use standard APIs like torch.empty to allocate
quantized tensors, with the understanding that we will set the
quantizer later.  This makes meta functions applicable to quantized
tensors (they will allocate with unknown quantizer and the kernel
will set the quantizer later) and fixes a bug David Dang reported
where structured kernels give a weird error message when you call them
with quantized inputs.

This is not a complete support for quantized structured kernels because
I haven't actually tried porting any of the quantized implementations
to structured; qadd is probably a good choice to try first as it
does its broadcasting implementation using TensorIterator.  My goal
here is just to show that the error message is better.

See also https://github.com/pytorch/pytorch/issues/52680

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D35317441](https://our.internmc.facebook.com/intern/diff/D35317441)

ezyang committed:
Towards supporting quantized structured kernels

This PR add support for quantized tensors with "unknown quantizer",
which means that we can use standard APIs like torch.empty to allocate
quantized tensors, with the understanding that we will set the
quantizer later.  This makes meta functions applicable to quantized
tensors (they will allocate with unknown quantizer and the kernel
will set the quantizer later) and fixes a bug David Dang reported
where structured kernels give a weird error message when you call them
with quantized inputs.

This is not a complete support for quantized structured kernels because
I haven't actually tried porting any of the quantized implementations
to structured; qadd is probably a good choice to try first as it
does its broadcasting implementation using TensorIterator.  My goal
here is just to show that the error message is better.

See also https://github.com/pytorch/pytorch/issues/52680

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bdhirsh commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -60,30 +60,15 @@ def gen_create_out_helper(backend_index: BackendIndex) -> List[str]:
-        runtime_empty_supported_check = ""
-    elif backend_index.dispatch_key == DispatchKey.CompositeExplicitAutograd:
+    elif backend_index.dispatch_key in (
+            DispatchKey.CompositeExplicitAutograd, DispatchKey.QuantizedCPU, DispatchKey.QuantizedCUDA):
```
tiny nit if we care about the perf (probably not necessary, just calling it out): we could avoid the dispatcher hop here from calling `at::empty`, but we'd have to abide by the naming convention and name the native kernels `at::native::empty_quantizedcpu/cuda`

ezyang replied:
oh this used to not be easy to do but now it is easy

bdhirsh commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -60,30 +60,15 @@ def gen_create_out_helper(backend_index: BackendIndex) -> List[str]:
         empty_impl = "at::empty"
         empty_strided_impl = "at::empty_strided"
-        runtime_empty_supported_check = """\
-  if (!c10::detail::backend_supports_empty_operator(options)) {{
```
thanks for killing this :')

jerryzh168 commented on test/test_dispatch.py:
```diff
@@ -935,5 +935,20 @@ def test_defaultbackend_math(self):
                 r"Registration to both CompositeImplicitAutograd and CompositeExplicitAutograd is not allowed"):
             dispatcher.register(["CompositeExplicitAutograd", "CompositeImplicitAutograd"])
 
+    def test_quantized_structured_not_implemented(self):
```
can we add a test in https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_tensor.py#L142 for calling some methods (e.g. qscheme) on an unknown tensor as well?

ezyang replied:
sure

jerryzh168 commented on aten/src/ATen/quantized/Quantizer.h:
```diff
@@ -18,6 +18,23 @@
+ * quantizer will be.
+ */
+struct TORCH_API UnknownQuantizer : public Quantizer {
+  explicit UnknownQuantizer(ScalarType scalar_type)
```
is scalar_type going to be used in the quantization kernels? can we document this?

ezyang replied:
I made UnknownQuantizer accept scalar type because the Quantizer superclass requires it. It shouldn't be used by anything. I'm actually a little surprised you store these on the Quantizer because it seems like you should be able to get them from the tensor.

jerryzh168 replied:
I see, yeah we need `scalar_type` because quantizer (as the name implies) is created before a quantized tensor, and it is able to quantize a floating point tensor to a tensor with a particular scalar type

jerryzh168 reviewed (approved):
followups are non-blocking, will accept first

ezyang commented:
@pytorchbot merge this

dzdang committed:
Update on "Towards supporting quantized structured kernels"


This PR add support for quantized tensors with "unknown quantizer",
which means that we can use standard APIs like torch.empty to allocate
quantized tensors, with the understanding that we will set the
quantizer later.  This makes meta functions applicable to quantized
tensors (they will allocate with unknown quantizer and the kernel
will set the quantizer later) and fixes a bug David Dang reported
where structured kernels give a weird error message when you call them
with quantized inputs.

This is not a complete support for quantized structured kernels because
I haven't actually tried porting any of the quantized implementations
to structured; qadd is probably a good choice to try first as it
does its broadcasting implementation using TensorIterator.  My goal
here is just to show that the error message is better.

See also https://github.com/pytorch/pytorch/issues/52680

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35317441).
